analyzers/tests/SonarAnalyzer.Test/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs

```diff
@@ -47,6 +46,313 @@ public void InstanceReference_SetsNotNull_VB()
+    public void Invocation_SetsNotNull_OnLinqMethodsNullChecking(string method, string args = null)
+    {
+        var code = $$"""
+            using System;
+            using System.Linq;
+            using System.Collections.Generic;
+            public class Sample
+            {
+                public void Main(IEnumerable<int> collection)
+                {
+                    Tag("Before1", collection);
+                    _ = collection.{{method}}({{args}});                                                 // Call as extension
+                    Tag("After1", collection);
+
+                    collection = Untracked();
+                    Tag("Before2", collection);
+                    _ = Enumerable.{{method}}(collection{{(args is null ? string.Empty : "," + args)}}); // Call as normal static method
+                    Tag("After2", collection);
+                }
+
+                private IEnumerable<int> Untracked() => [ 1, 2, 3 ];
+
+                private static void Tag(string name, object arg) { }
+            }
+            """;
+        var validator = new SETestContext(code, AnalyzerLanguage.CSharp, Array.Empty<SymbolicCheck>()).Validator;
+        validator.ValidateContainsOperation(OperationKind.Invocation);
+        validator.TagValue("Before1").Should().BeNull();
+        validator.TagValue("After1").Should().HaveOnlyConstraint(ObjectConstraint.NotNull);
+        validator.TagValue("Before2").Should().BeNull();
+        validator.TagValue("After2").Should().HaveOnlyConstraint(ObjectConstraint.NotNull);
+    }
+
+    [DataTestMethod]
+    [DataRow("collection.Cast<short>()")]
+    [DataRow("collection.OfType<short>()")]
+    [DataRow("collection.ThenBy(x => x)")]
+    [DataRow("collection.ThenByDescending(x => x)")]
+    public void Invocation_SetsNotNull_OnLinqMethodsWithGenericParamNullChecking(string invocation)
+    {
+        var code = $$"""
+            using System;
+            using System.Linq;
+            using System.Collections.Generic;
+            public class Sample
+            {
+                public void Main(IOrderedEnumerable<int> collection)
+                {
+                    Tag("Before1", collection);
+                    _ = {{invocation}};
+                    Tag("After1", collection);
+                }
+
+                private static void Tag(string name, object arg) { }
+            }
+            """;
+        var validator = new SETestContext(code, AnalyzerLanguage.CSharp, Array.Empty<SymbolicCheck>()).Validator;
+        validator.ValidateContainsOperation(OperationKind.Invocation);
+        validator.TagValue("Before1").Should().BeNull();
+        validator.TagValue("After1").Should().HaveOnlyConstraint(ObjectConstraint.NotNull);
+    }
+
+    [DataTestMethod]
+    [DataRow("OrderBy", "x => x")]
+    [DataRow("OrderByDescending", "x => x")]
+    [DataRow("GroupBy", "x => x")]
+    [DataRow("AsEnumerable")]
+    [DataRow("ToLookup", "x => x, x => x")]
```

Comment:
I don't get this test. None of them is null-tolerant AFAICT (I didn't check)

Reply:
I report in the description of the PR what we discussed offline with @Tim-Pohlmann, also to let @pavel-mikula-sonarsource (ping) and future maintainers of this work know about the approach taken and the reasons behind it.
Hope that clarifies the approach.

Reply:
@antonioaversa you mentioned methods that throw when "the enumerable is used somehow". Do you have any examples of this?
I think the easy solution is to look at the docs instead of the implementation and do a quick test when in doubt.

Reply:
I'm under the assumption that the documentation's exception section refers to exceptions that are thrown when invoking the method (aka what you refer to as early).

Reply:
Yes, I think so. Either way, the list of methods not throwing exceptions is now limited to the single `AsEnumerable` method (that surely doesn't raise exception, since it returns `source` without any check, and doesn't have any exception documented). All other methods in `Enumerable` now learn `NotNull`. I think it's consistent with the "check the doc for exception" approach.